### ISSUE TITLE
clang-tidy

### ISSUE THREAD

**Issue by JohnsterID** [LoneGazebo/Community-Patch-DLL#9110]
For info if anyone wants to check it out a clang-tidy run, found an [extension](https://marketplace.visualstudio.com/items?itemName=VladimirPetko.RunClangTidy-01) to run clang-tidy to the Visual Studio error list without changing from v90 build tools. Using LLVM 14.0.6. [Error list export tool](https://marketplace.visualstudio.com/items?itemName=ShemeerNS.ErrorListManager2ELM2).

[ErrorList_20220814175626.xlsx](https://github.com/LoneGazebo/Community-Patch-DLL/files/9333053/ErrorList_20220814175626.xlsx)

.clang-tidy
```
---
Checks: "-*,
         performance-*,
         portability-*,
         clang-analyzer-*,
"
WarningsAsErrors: ''
HeaderFilterRegex: ''
FormatStyle: none

```

edit: wrong hyperlink to tools,

**Comment by IanE9** [comment]
Could you export the error list as an HTML so that they can be viewed in browsers?

**Comment by JohnsterID** [comment]
Latest run attached. Github doesn't allow direct HTML attachments. Alternate? html in zip below.

Added --extra-arg=-std=c++98 --header-filter=.* --system-headers, ended up with a few thousands for C:\Program Files\Microsoft SDKs\Windows\v6.0A\include\basetsd.h and C:\Program Files\Microsoft SDKs\Windows\v6.0A\include\winnt.h.


[ErrorList_aad6f333a2.xlsx](https://github.com/LoneGazebo/Community-Patch-DLL/files/9336884/ErrorList_aad6f333a2.xlsx)
[ErrorList_aad6f333a2.pdf](https://github.com/LoneGazebo/Community-Patch-DLL/files/9336909/ErrorList_aad6f333a2.pdf)
[ErrorList_aad6f333a2-html.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/9336912/ErrorList_aad6f333a2-html.zip)

**Comment by IanE9** [comment]
I recently made it so that `NULL` would be defined as `nullptr` in clang builds so that we would actually get warnings related to improper uses of `NULL`. It looks like that change has confused it with the std set to c++98. You could try bumping it to c++11 and see if the `nullptr` warnings are suppressed, or we could just restrict the ifdef responsible for this further so it doesn't try and do this when the compiler is configured to c++98.

https://github.com/LoneGazebo/Community-Patch-DLL/blob/aad6f333a251566ca63060e57a6080ced60555ec/CvGameCoreDLL_Expansion2/CvGameCoreDLLPCH.h#L116-L122

**Comment by JohnsterID** [comment]
I will run again to check. Clang grouped c++03 into c++98. https://clang.llvm.org/cxx_status.html
I only added it to avoid code modernisation suggestions, might be easier to suppress those in .clang-tidy as well.

**Comment by LoneGazebo** [comment]
aside from faster compile, does clang offer any in-game benefit? If not, not sure if this is necessary.

**Comment by IanE9** [comment]
> aside from faster compile, does clang offer any in-game benefit? If not, not sure if this is necessary.

While not strictly necessary I personally consider Clang support to be pretty invaluable.

I'm mostly interested in being able to use Clang for the better codegen and diagnostics but here's a bullet point of potential benefits:
* Codegen for modern systems
* Better warning diagnostics by default
* Compatibility with modern analysis tools (clang-tidy for instance)
* Language level C++11 and up features become available
* Better code sync with @Hello71's [Linux port](https://github.com/Hello71/Community-Patch-DLL/tree/linux)

Basically all of the issues I've recently opened have been found because of Clang's better diagnostics, and you can review these PRs for locations where potential errors have been found and fixed:
* https://github.com/LoneGazebo/Community-Patch-DLL/pull/9112
* https://github.com/LoneGazebo/Community-Patch-DLL/pull/9062

**Comment by JohnsterID** [comment]
From my point of view, the ability to use more modern tools and project standards on project that is restricted due to linking is quite interesting to watch.
The effort by @IanE9 can lead to a better in-game experience by reducing the chance of bugs. Using other tools enables more coverage to ensure quality and security of the code base.

clang-tidy on 2.6. I have expanded the .clang-tidy checks with c++03 limitation to see what it generates. It appears clang-analyzer is not captured with it.

[ErrorList_304af1f.xlsx](https://github.com/LoneGazebo/Community-Patch-DLL/files/9354973/ErrorList_304af1f.xlsx)
[ErrorList_304af1f-html.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/9354974/ErrorList_304af1f-html.zip)
[ErrorList_304af1f-pdf.zip](https://github.com/LoneGazebo/Community-Patch-DLL/files/9354975/ErrorList_304af1f-pdf.zip)

**Comment by JohnsterID** [comment]
Close for now. Potential for Github Marketplace automation.
From the last giant check in the previous comment, I want to highlight 7944 instances of magic numbers. :)

**Comment by RecursiveVision** [comment]
@LoneGazebo It identified many problem areas in the code and Clang doesn't have the compiler bug which Visual Studio does that makes it occasionally impossible to load a release save in debug config.
